sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java

```diff
@@ -0,0 +1,253 @@
+                    .doOnComplete(() -> reInstantiateOpenConnectionsPublisherAndSubscribe())
+                    .subscribeOn(CosmosSchedulers.OPEN_CONNECTIONS_BOUNDED_ELASTIC)
+                    .subscribe();
+        }
+    }
+
+    public ProactiveOpenConnectionsProcessor(final RntbdEndpoint.Provider endpointProvider) {
+        this(endpointProvider, null);
+    }
+
+    public synchronized OpenConnectionTask submitOpenConnectionTaskOutsideLoop(
+        String collectionRid,
+        URI serviceEndpoint,
+        Uri addressUri,
+        int minConnectionsRequiredForEndpoint) {
+
+        OpenConnectionTask openConnectionTask = new OpenConnectionTask(collectionRid, serviceEndpoint, addressUri, minConnectionsRequiredForEndpoint);
+        this.submitOpenConnectionTaskOutsideLoopInternal(openConnectionTask);
+
+        return openConnectionTask;
+    }
+
+    private synchronized void submitOpenConnectionTaskOutsideLoopInternal(OpenConnectionTask openConnectionTask) {
+        String addressUriAsString = openConnectionTask.getAddressUri().getURIAsString();
+
+        this.endpointsUnderMonitorMap.compute(addressUriAsString, (key, taskList) -> {
+            if (taskList == null) {
+                taskList = new ArrayList<>();
+            }
+
+            taskList.add(openConnectionTask);
+            if (taskList.size() == 1) {
+                openConnectionsTaskSink.tryEmitNext(openConnectionTask);
+                openConnectionsTaskSinkBackUp.tryEmitNext(openConnectionTask);
+            }
+
+            return taskList;
+        });
+
+        RntbdEndpoint endpoint = this.endpointProvider.createIfAbsent(
+            openConnectionTask.getServiceEndpoint(),
+            openConnectionTask.getAddressUri().getURI(),
+            this,
+            openConnectionTask.getMinConnectionsRequiredForEndpoint());
+
+        endpoint.setMinChannelsRequired(Math.max(openConnectionTask.getMinConnectionsRequiredForEndpoint(), endpoint.getMinChannelsRequired()));
+    }
+
+    // There are two major kind tasks will be submitted
+    // 1. Tasks from up caller -> in this case, before we enqueue a task we would want to check whether the endpoint has already in the map
+    // 2. Tasks from existing flow -> for each endpoint, each time we will only 1 connection, if the connection count < the mini connection requirements,then re-queue.
+    // for this case, then there is no need to validate whether the endpoint exists
+    private synchronized void submitOpenConnectionWithinLoopInternal(OpenConnectionTask openConnectionTask) {
+        openConnectionsTaskSink.tryEmitNext(openConnectionTask);
+        openConnectionsTaskSinkBackUp.tryEmitNext(openConnectionTask);
+    }
+
+    @Override
+    public void close() throws IOException {
+        completeSink(openConnectionsTaskSink);
+        completeSink(openConnectionsTaskSinkBackUp);
+    }
+
+    public synchronized Disposable getOpenConnectionsPublisher() {
```

Comment:
Same here, this is very complex logic in this method, and doing I/O, it can cause deadlocks. Not sure why all of these need to be synchronized. 
@xinlian12 thoughts here?

Reply:
Resolved it offline.